File: dotnet/Safe/SafeHtmlOutput.cs
using SastEval.Util;

namespace SastEval.Safe;

/// <summary>
/// True Negative for CWE-79: User input is sanitized via HtmlSanitizer before rendering.
/// SAST tools should NOT flag this as XSS.
/// </summary>
public static class SafeHtmlOutput
{
    public static IResult Handle(HttpContext context)
    {
        var userInput = context.Request.Query["content"].ToString();

        if (string.IsNullOrEmpty(userInput))
        {
            return Results.BadRequest("Missing 'content' query parameter");
        }

        // SAFE: Input is sanitized through a strict HtmlSanitizer before rendering
        var sanitized = SanitizerUtil.Sanitize(userInput);

        return Results.Content($"<html><body><div>{sanitized}</div></body></html>", "text/html");
    }
}

File: java/src/main/java/com/sasteval/safe/SafeHtmlOutput.java
package com.sasteval.safe;

import com.sasteval.util.SanitizerUtil;

import javax.servlet.ServletException;
import javax.servlet.http.HttpServlet;
import javax.servlet.http.HttpServletRequest;
import javax.servlet.http.HttpServletResponse;
import java.io.IOException;

/**
 * TRUE NEGATIVE for CWE-79: XSS
 * SAFE: User input is sanitized through SanitizerUtil before output.
 */
public class SafeHtmlOutput extends HttpServlet {

    @Override
    protected void doGet(HttpServletRequest request, HttpServletResponse response)
            throws ServletException, IOException {

        String userInput = request.getParameter("input");
        response.setContentType("text/html");

        // SAFE: Input is sanitized before rendering
        String sanitized = SanitizerUtil.sanitize(userInput);
        response.getWriter().println("<div>" + sanitized + "</div>");
    }
}
